load-test/bin/generate-pupil-load-test-data.js

```diff
@@ -1,101 +1,107 @@
-          value: name,
-          type: TYPES.NVarChar
-        },
-        {
-          name: 'urn',
-          value: urn,
-          type: TYPES.Int
-        },
-        {
-          name: 'dfeNumber',
-          value: dfeNumber,
-          type: TYPES.Int
-        }
-      ]
-      const sql = `INSERT INTO ${sqlService.adminSchema}.school
-      (leaCode, estabCode, name, urn, dfeNumber)
-      VALUES (999, @estabCode, @name, @urn, @dfeNumber)`
-      await sqlService.query(sql, params)
+
+    let schools, numSchools, pupilsPerSchool, pupilsRemainder, params
+
+    schools = await sqlService.query(`SELECT id FROM ${sqlService.adminSchema}.[school]`)
+    numSchools = schools.length
+
+    // When generating less pupils than the total number of schools, slice
+    // the schools array to the number of pupils and set 1 pupil per school
+    pupilsPerSchool = Math.floor(numPupils / numSchools)
+    if (pupilsPerSchool < 1) {
+      pupilsPerSchool = 1
+      schools = schools.slice(0, numPupils)
+      numSchools = schools.length
+      pupilsRemainder = 0
+    } else {
+      pupilsRemainder = numPupils - (pupilsPerSchool * numSchools)
     }
-    const testSchools = await schoolDataService.sqlFindByDfeNumbers(dfeNumbers)
-    // Generate additional pupils for these schools and pins
-    winston.info('Generating pupils within schools and pins...')
-    await Promise.all(testSchools.map(async s => {
-      const params = [
+
+    winston.info(`Generating ${numPupils} pupils across ${numSchools} schools`)
+
+    for (let i = 0; i < numSchools; i++) {
+      let school = schools[i]
+      let totalPupils = pupilsPerSchool
+      if (i < pupilsRemainder) {
+        totalPupils += 1
+      }
+      params = [
         {
           name: 'schoolId',
-          value: s.id,
-          type: TYPES.NVarChar
+          value: school.id,
+          type: TYPES.Int
         },
         {
           name: 'dateOfBirth',
           value: randomDob(),
           type: TYPES.DateTimeOffset
+        },
+        {
+          name: 'pinExpiresAt',
+          value: pinExpireyDate,
+          type: TYPES.DateTimeOffset
         }
       ]
+
       const sql = `
-      DECLARE @cnt INT = 0;
+      DECLARE @cnt INT = 1;
       DECLARE @baseUpn INT = 80120000 + @schoolId
-      WHILE @cnt < 60
+      BEGIN TRAN
+
+      UPDATE ${sqlService.adminSchema}.[school]
+      SET pin = '${randomPass()}', pinExpiresAt = @pinExpiresAt
+      WHERE id = ${school.id} AND pin IS NULL
+
+      WHILE @cnt <= ${totalPupils}
       BEGIN
-        INSERT ${sqlService.adminSchema}.[pupil] (school_id, foreName, lastName, gender, dateOfBirth, upn) 
-        VALUES (@schoolId, CAST(@cnt AS NVARCHAR), 'Pupil', 'M', @dateOfBirth, CAST(@baseUpn AS NVARCHAR) + CAST(@cnt AS NVARCHAR) + '1A')
-        SET @cnt = @cnt + 1;
-      END;`
+      BEGIN TRY
+
+      INSERT ${sqlService.adminSchema}.[pupil] (school_id, foreName, lastName, gender, dateOfBirth, upn, isTestAccount, pin, pinExpiresAt) 
```

Comment:
There seems little point in populating the `pupil`.`pin` col - it is only used by the legacy code.  Load testing needs to test the HPA architecture, and PR #867 is moving the HPA pin from `check`.`pin` a new table `checkPin` which joins to `pin`.

In fact, once PR #867 is merged there will be a new stored procedure to generate checks and pins.  I'd suggest waiting for that.

Reply:
we appear to have caught up in a short space of time.
@danwilliams one thing i overlooked in our conversation yesterday was the fact that this piece of work is specifically for the HPA!